File: app/src/main/java/org/wen/gank/widgets/NestedWebView.java
package org.wen.gank.widgets;

import android.content.Context;
import android.support.v4.view.MotionEventCompat;
import android.support.v4.view.NestedScrollingChild;
import android.support.v4.view.NestedScrollingChildHelper;
import android.support.v4.view.ViewCompat;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.webkit.WebView;

/**
 * created by Jiahui.wen 2017-09-20
 */
public class NestedWebView extends WebView implements NestedScrollingChild {

    private int mLastY;
    private final int[] mScrollOffset = new int[2];
    private final int[] mScrollConsumed = new int[2];

    private NestedScrollingChildHelper nestedChildHelper;

    public NestedWebView(Context context) {
        this(context, null);
    }

    public NestedWebView(Context context, AttributeSet attrs) {
        this(context, attrs, android.R.attr.webViewStyle);
    }

    public NestedWebView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        nestedChildHelper = new NestedScrollingChildHelper(this);
        setNestedScrollingEnabled(true);
    }

    @Override
    public boolean onTouchEvent(MotionEvent ev) {
        boolean returnValue = false;

        MotionEvent event = MotionEvent.obtain(ev);
        final int action = MotionEventCompat.getActionMasked(event);

        int eventY = (int) event.getY();

        switch (action) {
            case MotionEvent.ACTION_MOVE:
                int deltaY = mLastY - eventY;
                if (dispatchNestedPreScroll(0, deltaY, mScrollConsumed, mScrollOffset)) {
                    deltaY -= mScrollConsumed[1];
                }
                returnValue = super.onTouchEvent(event);
                if (dispatchNestedScroll(0, mScrollOffset[1], 0, deltaY, mScrollOffset)) {
                    mLastY -= mScrollOffset[1];
                }
                break;
            case MotionEvent.ACTION_DOWN:
                mLastY = eventY;
                returnValue = super.onTouchEvent(event);
                startNestedScroll(ViewCompat.SCROLL_AXIS_VERTICAL);
                break;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                returnValue = super.onTouchEvent(event);
                stopNestedScroll();
                break;
        }
        return returnValue;
    }

    @Override
    public void setNestedScrollingEnabled(boolean enabled) {
        nestedChildHelper.setNestedScrollingEnabled(enabled);
    }

    @Override
    public boolean isNestedScrollingEnabled() {
        return nestedChildHelper.isNestedScrollingEnabled();
    }

    @Override
    public boolean startNestedScroll(int axes) {
        return nestedChildHelper.startNestedScroll(axes);
    }

    @Override
    public void stopNestedScroll() {
        nestedChildHelper.stopNestedScroll();
    }

    @Override
    public boolean hasNestedScrollingParent() {
        return nestedChildHelper.hasNestedScrollingParent();
    }

    @Override
    public boolean dispatchNestedScroll(int dxConsumed, int dyConsumed, int dxUnconsumed,
        int dyUnconsumed, int[] offsetInWindow) {
        return nestedChildHelper.dispatchNestedScroll(dxConsumed, dyConsumed, dxUnconsumed,
            dyUnconsumed, offsetInWindow);
    }

    @Override
    public boolean dispatchNestedPreScroll(int dx, int dy, int[] consumed, int[] offsetInWindow) {
        return nestedChildHelper.dispatchNestedPreScroll(dx, dy, consumed, offsetInWindow);
    }

    @Override
    public boolean dispatchNestedFling(float velocityX, float velocityY, boolean consumed) {
        return nestedChildHelper.dispatchNestedFling(velocityX, velocityY, consumed);
    }

    @Override
    public boolean dispatchNestedPreFling(float velocityX, float velocityY) {
        return nestedChildHelper.dispatchNestedPreFling(velocityX, velocityY);
    }
}

File: app/src/main/java/org/wen/gank/widgets/NestedWebView.kt
package org.wen.gank.widgets

import android.content.Context
import android.support.v4.view.NestedScrollingChild
import android.support.v4.view.NestedScrollingChildHelper
import android.support.v4.view.ViewCompat
import android.util.AttributeSet
import android.view.MotionEvent
import android.webkit.WebView

/**
 * created by Jiahui.wen 2017-09-20
 */
class NestedWebView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = android.R.attr.webViewStyle) : WebView(context, attrs, defStyleAttr), NestedScrollingChild {

  private var mLastY: Int = 0
  private val mScrollOffset = IntArray(2)
  private val mScrollConsumed = IntArray(2)

  private val nestedChildHelper: NestedScrollingChildHelper = NestedScrollingChildHelper(this)

  init {
    isNestedScrollingEnabled = true
  }

  override fun onTouchEvent(ev: MotionEvent): Boolean {
    var returnValue = false

    val event = MotionEvent.obtain(ev)
    val action = event.actionMasked

    val eventY = event.y.toInt()

    when (action) {
      MotionEvent.ACTION_MOVE -> {
        var deltaY = mLastY - eventY
        if (dispatchNestedPreScroll(0, deltaY, mScrollConsumed, mScrollOffset)) {
          deltaY -= mScrollConsumed[1]
        }
        returnValue = super.onTouchEvent(event)
        if (dispatchNestedScroll(0, mScrollOffset[1], 0, deltaY, mScrollOffset)) {
          mLastY -= mScrollOffset[1]
        }
      }
      MotionEvent.ACTION_DOWN -> {
        mLastY = eventY
        returnValue = super.onTouchEvent(event)
        startNestedScroll(ViewCompat.SCROLL_AXIS_VERTICAL)
      }
      MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
        returnValue = super.onTouchEvent(event)
        stopNestedScroll()
      }
    }
    return returnValue
  }

  override fun setNestedScrollingEnabled(enabled: Boolean) {
    nestedChildHelper.isNestedScrollingEnabled = enabled
  }

  override fun isNestedScrollingEnabled(): Boolean {
    return nestedChildHelper.isNestedScrollingEnabled
  }

  override fun startNestedScroll(axes: Int): Boolean {
    return nestedChildHelper.startNestedScroll(axes)
  }

  override fun stopNestedScroll() {
    nestedChildHelper.stopNestedScroll()
  }

  override fun hasNestedScrollingParent(): Boolean {
    return nestedChildHelper.hasNestedScrollingParent()
  }

  override fun dispatchNestedScroll(dxConsumed: Int, dyConsumed: Int, dxUnconsumed: Int,
                                    dyUnconsumed: Int, offsetInWindow: IntArray?): Boolean {
    return nestedChildHelper.dispatchNestedScroll(dxConsumed, dyConsumed, dxUnconsumed,
        dyUnconsumed, offsetInWindow)
  }

  override fun dispatchNestedPreScroll(dx: Int, dy: Int, consumed: IntArray?, offsetInWindow: IntArray?): Boolean {
    return nestedChildHelper.dispatchNestedPreScroll(dx, dy, consumed, offsetInWindow)
  }

  override fun dispatchNestedFling(velocityX: Float, velocityY: Float, consumed: Boolean): Boolean {
    return nestedChildHelper.dispatchNestedFling(velocityX, velocityY, consumed)
  }

  override fun dispatchNestedPreFling(velocityX: Float, velocityY: Float): Boolean {
    return nestedChildHelper.dispatchNestedPreFling(velocityX, velocityY)
  }
}
